pipeat.go

```diff
@@ -32,34 +32,65 @@ func (c spans) Less(i, j int) bool { return c[i].start < c[j].start }
 type pipeFile struct {
 	*os.File
 	fileLock  sync.RWMutex
-	sync.Cond              // used to signal readers from writers
+	rCond     sync.Cond    // used to signal readers from writers
+	wCond     sync.Cond    // used to signal writers from readers in sync mode
 	dataLock  sync.RWMutex // serialize access to meta data (below)
+	syncMode  bool         // if true the writer is only allowed to write until the reader requested
 	endln     int64
 	ahead     spans
 	readeroff int64 // offset where Read() last read
+	writeroff int64 // file offset allowed for the writer in sync mode
+	readed    int64 // size readed as bytes, useful for stats
+	written   int64 // size written as bytes, useful for stats
 	rerr      error
 	werr      error
 	eow       chan struct{} // end of writing
 	eor       chan struct{} // end of reading
 }
 
-func newPipeFile() (*pipeFile, error) {
-	file, err := ioutil.TempFile("", "pipefile")
+func newPipeFile(dirPath string) (*pipeFile, error) {
+	file, err := ioutil.TempFile(dirPath, "pipefile")
+	if err != nil {
+		return nil, err
+	}
+	file, err = unlinkFile(file)
 	if err != nil {
 		return nil, err
 	}
-	os.Remove(file.Name())
 	f := &pipeFile{File: file,
 		eow: make(chan struct{}),
 		eor: make(chan struct{})}
-	f.L = f.fileLock.RLocker() // Cond locker
+	f.rCond.L = f.dataLock.RLocker() // Readers cond locker
+	f.wCond.L = f.dataLock.RLocker() // Writers cond locker
 	return f, nil
 }
 
-func (f *pipeFile) readable(start, end int64) bool {
+// waitForReadable waits until the requested data is available to read.
+// We can stop waiting if the writer ended or if the reader is closed.
+// The reader can be closed using Close() or CloseWithError we don't
+// want to wait on a closed reader
+func (f *pipeFile) waitForReadable(off int64) {
 	f.dataLock.RLock()
 	defer f.dataLock.RUnlock()
-	return (start+end < f.endln)
+L:
+	for off >= f.endln {
+		select {
+		case <-f.eow:
+			trace("eow")
+			break L
```

Comment:
Style preference.... I prefer to use returns when possible instead of breaks. Could you change the use of breaks like this to returns?